consensus/safety-rules/src/safety_rules.rs

```diff
@@ -89,17 +104,53 @@ impl SafetyRules {
         }
     }
 
+    /// This reconciles the key pair of a validator signer with a given validator set
+    /// during epoch changes.
+    /// @TODO Given we cannot panic, we must handle the following two error cases:
+    ///     1. Validator not in the set
+    ///     2. Validator in the set, but no matching key found in storage
```

Comment:
this comment is strange, we should never panic, I don't know why we would have this comment here or written like this. Perhaps you want to say we shouldn't make consensus panic, but I think that is actually okay.

Reply:
It was me suggesting we should just panic if the key not found but then realize we can't

Reply:
Just fixed the comment here. Does that make sense now?